lib/Authentication.ts

```diff
@@ -49,6 +52,150 @@ export default class Authentication {
+      throw new Error('Signing DID does not match issuer');
+    }
+    return verifiedRequest;
+  }
+
+  /**
+   * Given a challenge, forms a signed response using a given DID that expires at expiration, or a default expiration.
+   * @param authRequest Challenge to respond to
+   * @param responseDid The DID to respond with
+   * @param claims Claims that the requester asked for
+   * @param expiration optional expiration datetime of the response
+   */
+  public async formAuthenticationResponse (authRequest: AuthenticationRequest, responseDid: string, claims: any, expiration?: Date): Promise<string> {
+    let key: PrivateKey | undefined;
+    key = this.getKey(responseDid);
+    if (!key) {
+      throw new Error(`Could not find a key for ${responseDid}`);
+    }
+
+    const publicKey: PublicKey = key.getPublicKey();
+
+    // milliseconds to seconds
+    const milliseconds = 1000;
+    if (!expiration) {
+      const expirationTimeOffsetInMinutes = 5;
+      expiration = new Date(Date.now() + milliseconds * 60 * expirationTimeOffsetInMinutes); // 5 minutes from now
+    }
+    const iat = Math.floor(Date.now() / milliseconds); // ms to seconds
+    const response: AuthenticationResponse = {
+      iss: 'https://self-issued.me',
+      sub: responseDid,
+      aud: authRequest.client_id,
+      nonce: authRequest.nonce,
+      exp: Math.floor(expiration.getTime() / milliseconds),
+      iat,
+      sub_jwk: {},
+      did: responseDid,
+      state: authRequest.state
+    };
+
+    let authResponse = Object.assign(response, { sub_jwk: publicKey });
+    authResponse = Object.assign(authResponse, claims);
+
+    const token = new JwsToken(authResponse, this.factory);
+    return token.sign(key, {
+      iat: iat.toString(),
+      exp: Math.floor(expiration.getTime() / milliseconds).toString()
+    });
+  }
+
+  /**
+   * Private method that gets the private key of the DID from the key mapping.
+   * @param did the DID whose private key is used to sign JWT.
+   */
+  private getKey (did: string): PrivateKey | undefined {
+    let key: PrivateKey | undefined;
+    for (const keyId in this.keys) {
+      if (keyId.startsWith(did)) {
+        key = this.keys[keyId];
+        break;
+      }
+    }
+    return key;
+  }
+
+  /**
+   * Verifies the signature on a AuthenticationResponse and returns a AuthenticationResponse object
+   * @param authResponse AuthenticationResponse to verify as a string or buffer
+   * @returns the authenticationResponse as a AuthenticationResponse
+   */
+  public async verifyAuthenticationResponse (authResponse: Buffer | string): Promise<AuthenticationResponse> {
+    const clockSkew = 5 * 60 * 1000; // 5 minutes
+    let jwsToken: JwsToken;
+    if (authResponse instanceof Buffer) {
+      jwsToken = new JwsToken(authResponse.toString(), this.factory);
+    } else {
+      jwsToken = new JwsToken(authResponse, this.factory);
+    }
+    const exp = jwsToken.getHeader().exp;

```

Comment:
What happens if the **_exp_** or one of the other required JWT properties isn't provided? Does the JwsToken constructor check for the required properties? If not there should be a check and any tokens that don't conform a rejected out of hand.

Reply:
Is `exp` required? Theres a check below to enforce the expiration if it exists.

Reply:
According to [RFC 7519 4.1.4 "exp" Claim](https://tools.ietf.org/html/rfc7519#section-4.1.4), it's optional.